activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -303,6 +303,151 @@ def _select!(*fields) # :nodoc:
       self
     end
 
+    # Add a Common Table Expression (CTE) that you can then reference within another SELECT statement.
+    #
+    #   Post.with(:posts_with_tags, Post.where("tags_count > ?", 0))
+    #   # WITH posts_with_tags AS (
+    #   #   SELECT * FROM posts WHERE (tags_count > 0)
+    #   # )
+    #   # SELECT * FROM posts
+    #
+    # Once you define Common Table Expression you can use custom `FROM` value or `JOIN` to reference it.
+    #
+    #   Post.with(:posts_with_tags, Post.where("tags_count > ?", 0).from("posts_with_tags AS posts")
+    #   # WITH posts_with_tags AS (
+    #   #  SELECT * FROM posts WHERE (tags_count > 0)
+    #   # )
+    #   # SELECT * FROM posts_with_tags AS posts
+    #
+    #   Post.with(:posts_with_tags, Post.where("tags_count > ?", 0).joins("JOIN posts_with_tags ON posts_with_tags.id = posts.id"))
```

Comment:
I think the parens on this are wrong. Maybe:

```suggestion
    #   Post.with(:posts_with_tags, Post.where("tags_count > ?", 0)).joins("JOIN posts_with_tags ON posts_with_tags.id = posts.id")
```

B. It is too bad we have to resort to this joins clause.
I say this because it feels like I will want to join the primary model to the CTE for most (all?) of my uses of CTE. Not sure if it is the same for you.

Is it possible to implement:

```
Post.with(:posts_with_tags, Post.where("tags_count > ?", 0)).joins(:posts_with_tags)
```

Reply:
I think it is possible and I proposed some ideas already https://github.com/rails/rails/pull/37944#issuecomment-565847607 but as suggested I wouldn't implement it here to keep this PR small and easy to grasp.

Simple scopes like [this](https://github.com/rails/rails/pull/37944/files#diff-f61c8d8a428f9dbae2dc668ca71f894e6f2b5947fb28a6a9339bf2854f89073aR64) could be enough for start maybe.